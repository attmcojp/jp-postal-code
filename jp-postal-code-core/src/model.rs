/// 郵便番号レコード
///
/// オリジナルの情報を残すようなモデルにしている。
///
/// # Example
/// ```rust
/// # use jp_postal_code_core::model::UtfKenAllRecord;
/// let _ = UtfKenAllRecord {
///     local_government_code: "01101".to_string(),
///     old_postal_code: "060  ".to_string(),
///     postal_code: "0600000".to_string(),
///     prefecture_kana: "ホッカイドウ".to_string(),
///     city_kana: "サッポロシチュウオウク".to_string(),
///     town_kana: "イカニケイサイガナイバアイ".to_string(),
///     prefecture: "北海道".to_string(),
///     city: "札幌市中央区".to_string(),
///     town: "以下に掲載がない場合".to_string(),
///     has_multi_postal_code: 0,
///     has_chome: 0,
///     has_multi_town: 0,
///     update_code: 0,
///     update_reason: 0,
/// };
/// ```
///
/// # References
/// - [郵便番号データ（1レコード1行、UTF-8形式）の説明](https://www.post.japanpost.jp/zipcode/dl/utf-readme.html)
///
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
pub struct UtfKenAllRecord {
    /// 全国地方公共団体コード（JIS X0401、X0402）
    pub local_government_code: String,
    /// 旧郵便番号（5桁）
    pub old_postal_code: String,
    /// 郵便番号（7桁）
    pub postal_code: String,
    /// 都道府県名（カタカナ）
    pub prefecture_kana: String,
    /// 市区町村名（カタカナ）
    pub city_kana: String,
    /// 町域名（カタカナ）
    pub town_kana: String,
    /// 都道府県名
    pub prefecture: String,
    /// 市区町村名
    pub city: String,
    /// 町域名
    pub town: String,
    /// 一町域が二以上の郵便番号で表される場合の表示
    ///
    /// - 0: 該当せず
    /// - 1: 該当
    pub has_multi_postal_code: i16,
    /// 丁目を有する町域の場合の表示
    ///
    /// - 0: 該当せず
    /// - 1: 該当
    pub has_chome: i16,
    /// 一つの郵便番号で二以上の町域を表す場合の表示
    ///
    /// - 0: 該当せず
    /// - 1: 該当
    pub has_multi_town: i16,
    /// 更新の表示
    ///
    /// - 0: 変更なし
    /// - 1: 変更あり
    /// - 2: 廃止
    pub update_code: i16,
    /// 更新理由
    ///
    /// - 0: 変更なし
    /// - 1: 市政・区政・町政・分区・政令指定都市施行
    /// - 2: 住居表示の実施
    /// - 3: 区画整理
    /// - 4: 郵便区調整等
    /// - 5: 訂正
    /// - 6: 廃止
    pub update_reason: i16,
}
