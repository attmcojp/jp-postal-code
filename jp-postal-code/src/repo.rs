use jp_postal_code_core::model::UtfKenAllRecord;

pub const DEFAULT_SEARCH_PAGE_SIZE: usize = 10;

#[derive(Debug, Clone)]
pub struct UtfKenAllRepositorySearchRequest<'a> {
    pub postal_code: &'a str,
    pub page_size: Option<usize>,
    pub page_token: Option<&'a str>,
}

#[derive(Debug, Clone)]
pub struct UtfKenAllRepositorySearchResponse {
    pub records: Vec<UtfKenAllRecord>,
    pub next_page_token: Option<String>,
}

/// 郵便番号データベースを扱うリポジトリ
pub trait UtfKenAllRepository: Clone + Send + Sync {
    type Error: std::error::Error + Send + Sync + 'static;

    /// 郵便番号データベースを置き換える
    fn replace(
        &mut self,
        records: &[UtfKenAllRecord],
    ) -> impl std::future::Future<Output = Result<(), Self::Error>> + Send;

    /// 郵便番号データベースから前方一致でレコードを検索する
    fn search(
        &self,
        req: UtfKenAllRepositorySearchRequest<'_>,
    ) -> impl std::future::Future<Output = Result<UtfKenAllRepositorySearchResponse, Self::Error>> + Send;

    /// 郵便番号データベースの総数をカウントする
    fn count(&self) -> impl std::future::Future<Output = Result<usize, Self::Error>> + Send;
}
