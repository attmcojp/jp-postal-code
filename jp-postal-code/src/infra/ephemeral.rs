use crate::repo::UtfKenAllRepository;
use jp_postal_code_core::model::UtfKenAllRecord;
use std::sync::{Arc, Mutex};

#[derive(Debug, Default, Clone)]
pub struct UtfKenAllRepositoryEphemeral {
    records: Arc<Mutex<Vec<UtfKenAllRecord>>>,
}

impl UtfKenAllRepositoryEphemeral {
    pub fn new(records: Vec<UtfKenAllRecord>) -> Self {
        Self {
            records: Arc::new(Mutex::new(records)),
        }
    }

    pub fn into_inner(self) -> Arc<Mutex<Vec<UtfKenAllRecord>>> {
        self.records
    }
}

impl UtfKenAllRepository for UtfKenAllRepositoryEphemeral {
    type Error = std::convert::Infallible;

    #[tracing::instrument(skip(self, records))]
    async fn replace(&mut self, records: &[UtfKenAllRecord]) -> Result<(), Self::Error> {
        self.records
            .lock()
            .unwrap()
            .splice(0.., records.iter().cloned());
        Ok(())
    }

    #[tracing::instrument(skip(self))]
    async fn search(&self, postal_code: &str) -> Result<Vec<UtfKenAllRecord>, Self::Error> {
        let mut found = Vec::new();
        self.records.lock().unwrap().iter().for_each(|r| {
            if r.postal_code.starts_with(postal_code) {
                found.push(r.clone())
            }
        });
        Ok(found)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn utf_ken_all_repository_ephemeral_replace() {
        let mut repository = UtfKenAllRepositoryEphemeral::default();

        // 正しくデータが挿入されているかチェック
        repository
            .replace(&[
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "060  ".to_string(),
                    postal_code: "0600000".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "イカニケイサイガナイバアイ".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "以下に掲載がない場合".to_string(),
                    has_multi_postal_code: 0,
                    has_chome: 0,
                    has_multi_town: 0,
                    update_code: 0,
                    update_reason: 0,
                },
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "064  ".to_string(),
                    postal_code: "0640941".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "アサヒガオカ".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "旭ケ丘".to_string(),
                    has_multi_postal_code: 0,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "060  ".to_string(),
                    postal_code: "0600041".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "オオドオリヒガシ".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "大通東".to_string(),
                    has_multi_postal_code: 0,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "060  ".to_string(),
                    postal_code: "0600042".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "オオドオリニシ（１−１９チョウメ）".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "大通西（１〜１９丁目）".to_string(),
                    has_multi_postal_code: 1,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "064  ".to_string(),
                    postal_code: "0640820".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "オオドオリニシ（２０−２８チョウメ）".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "大通西（２０〜２８丁目）".to_string(),
                    has_multi_postal_code: 1,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
            ])
            .await
            .unwrap();
        insta::assert_debug_snapshot!(repository.records.lock().unwrap());

        // 正しく置き換えられるかチェック
        repository
            .replace(&[
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "064  ".to_string(),
                    postal_code: "0640941".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "アサヒガオカ".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "旭ケ丘".to_string(),
                    has_multi_postal_code: 0,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "060  ".to_string(),
                    postal_code: "0600042".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "オオドオリニシ（１−１９チョウメ）".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "大通西（１〜１９丁目）".to_string(),
                    has_multi_postal_code: 1,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
            ])
            .await
            .unwrap();
        insta::assert_debug_snapshot!(repository.records.lock().unwrap());
    }

    #[tokio::test]
    async fn utf_ken_all_repository_ephemeral_search() {
        let repository = UtfKenAllRepositoryEphemeral::new(vec![
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600000".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "イカニケイサイガナイバアイ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "以下に掲載がない場合".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 0,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "064  ".to_string(),
                postal_code: "0640941".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "アサヒガオカ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "旭ケ丘".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600041".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリヒガシ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通東".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600042".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリニシ（１−１９チョウメ）".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通西（１〜１９丁目）".to_string(),
                has_multi_postal_code: 1,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "064  ".to_string(),
                postal_code: "0640820".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリニシ（２０−２８チョウメ）".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通西（２０〜２８丁目）".to_string(),
                has_multi_postal_code: 1,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
        ]);

        // 完全一致で検索
        let records = repository.search("0640820").await.unwrap();
        insta::assert_debug_snapshot!(records);

        // 前方一致で検索
        let records = repository.search("060").await.unwrap();
        insta::assert_debug_snapshot!(records);
    }
}
