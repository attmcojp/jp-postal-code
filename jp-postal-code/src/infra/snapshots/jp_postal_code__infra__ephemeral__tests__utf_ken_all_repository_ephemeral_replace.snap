---
source: jp-postal-code/src/infra/ephemeral.rs
expression: repository.records
---
[
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600000",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "イカニケイサイガナイバアイ",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "以下に掲載がない場合",
        has_multi_postal_code: 0,
        has_chome: 0,
        has_multi_town: 0,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640941",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "アサヒガオカ",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "旭ケ丘",
        has_multi_postal_code: 0,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600041",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリヒガシ",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通東",
        has_multi_postal_code: 0,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "060  ",
        postal_code: "0600042",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（１−１９チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西（１〜１９丁目）",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640820",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（２０−２８チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西（２０〜２８丁目）",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
]
